game-core/src/main/java/games/strategy/engine/lobby/server/login/LobbyLoginValidator.java

```diff
@@ -74,14 +75,18 @@
   private final RsaAuthenticator rsaAuthenticator;
   private final UserDao userDao;
 
-  public LobbyLoginValidator() {
+  public LobbyLoginValidator(final LobbyPropertyReader lobbyPropertyReader) {
+    this(lobbyPropertyReader, new Database(lobbyPropertyReader));
```

Comment:
Would it make sense to go ahead and inject the `Database` instance?

Reply:
I don't think so.  That would create a dependency on the caller to the persistence layer (via `g.s.engine.lobby.server.db`).  The caller would essentially be skipping a layer in the architecture.

Note that this funky-looking code was only added to work around the limitations of constructor chaining.  I only introduced the new private constructor to avoid the public constructor creating numerous `Database` instances.  For example:

```java
  public LobbyLoginValidator(final LobbyPropertyReader lobbyPropertyReader) {
    this(
        lobbyPropertyReader,
        new BadWordController(new Database(lobbyPropertyReader)),
        new BannedMacController(new Database(lobbyPropertyReader)),
        new BannedUsernameController(new Database(lobbyPropertyReader)),
        new UserController(new Database(lobbyPropertyReader)),
        new CompositeAccessLog(new AccessLogController(new Database(lobbyPropertyReader))),
        new RsaAuthenticator(),
        BCrypt::gensalt);
  }
```

The multiple `Database` instances are not a big deal since they're extremely lightweight, but it just felt dirty. :smile:

Reply:
I don't think so.  That would create a dependency from the caller to the persistence layer (via `g.s.engine.lobby.server.db`).  The caller would essentially be skipping a layer in the architecture.

Note that this funky-looking code was only added to work around the limitations of constructor chaining.  I only introduced the new private constructor to avoid the public constructor creating numerous `Database` instances.  For example:

```java
  public LobbyLoginValidator(final LobbyPropertyReader lobbyPropertyReader) {
    this(
        lobbyPropertyReader,
        new BadWordController(new Database(lobbyPropertyReader)),
        new BannedMacController(new Database(lobbyPropertyReader)),
        new BannedUsernameController(new Database(lobbyPropertyReader)),
        new UserController(new Database(lobbyPropertyReader)),
        new CompositeAccessLog(new AccessLogController(new Database(lobbyPropertyReader))),
        new RsaAuthenticator(),
        BCrypt::gensalt);
  }
```

The multiple `Database` instances are not a big deal since they're extremely lightweight, but it just felt dirty. :smile:

Reply:
If you go up one more layer though?
```
  public LobbyLoginValidator(final Database database) {
    this(
        lobbyPropertyReader,
        new BadWordController(database),
        new BannedMacController(database),
        new BannedUsernameController(database),
        new UserController(database),
        new CompositeAccessLog(new AccessLogController(database)),
        new RsaAuthenticator(),
        BCrypt::gensalt);
  }
```

If controller should not know about getting DB, then it should get a DB supplier. Typically you're going to have a container create a threadpool with DB connections, a handle to that is then injected. 

In terms of DI there is a still a smell here:
```
this(lobbyPropertyReader, new Database(lobbyPropertyReader));
```
The property reader is injected but we new up Database with the same injected parameter.

Reply:
Again, that pushes knowledge of `Database` to the owner of `LobbyLoginValidator`, which doesn't seem appropriate.  I view `LobbyLoginValidator` as somewhere in the service layer, and it's API probably shouldn't contain types from the `db` package.

> If controller should not know about getting DB, then it should get a DB supplier.

It's perfectly fine for the _controllers_ (DAOs) to depend on `Database` since they're already in the same package.  My objection was to having `LobbyServer` depend on `Database`.

> Typically you're going to have a container create a threadpool with DB connections, a handle to that is then injected.

Note that, in its current state, `Database` is effectively a `Supplier<Connection>`.

> Hence, I think a lot of things are going to made easier if the DB params or DB itself is injected.

The DB params _are_ being injected via `LobbyPropertyReader` (from the higher layer) and converted to `Database` (for the lower layer).  We're effectively doing a translation of the DB params between domains when crossing through `LobbyLoginValidator` in order to not pollute APIs facing higher layers with lower layer types that the higher layers don't need to be concerned about.

To take this separation of domains to the extreme, one could argue that `Database` shouldn't depend on `LobbyPropertyReader` and that the individual parameters should be specified instead (possibly via a _Parameter Object_).  However, I took the lazy way out since that dependency already existed. :smile:

Reply:
Hmm, I'll try to clarify the concern/suggestion, injecting something to read file system to get DB creds which can then be used to create a DB connections seems very round-about compared to injecting the DB connection directly, ie: exactly a `Supplier<Connection>`. 

As pseudo-code, I kinda view it as:
```
public AnotherController(PropertyReader reader) {
	Database db = new Database(reader.readProps);
}

```
When it could be:
```
public AnotherController(Database db) {
   :
}

```

From a test perspective, I know I'd prefer the latter so the DB could be mocked or stubbed.



> Again, that pushes knowledge of Database to the owner of LobbyLoginValidator

I view that as a good thing as I don't think `LobbyLoginValidator` should know the details of how to `new` up a DB conn just as `BadWordController` should not know how to `new` up a DB conn either. 


>  I view LobbyLoginValidator as somewhere in the service layer, and it's API probably shouldn't contain types from the db package.

I would tend to agree. Currently it needs to know how to instantiate it's dependendencies, notably the controllers, which need a DB conn. That means in the current setup `LobbyLoginValidator` one way or another needs to know how to create a DB conn.

One way to solve that would be to inject the controllers into `LobbyLoginValidator`.  That is funn enough effectively what is going on when we have:
```

  public LobbyLoginValidator(final Database database) {
    this(
        lobbyPropertyReader,
        new BadWordController(database),
        new BannedMacController(database),
        new BannedUsernameController(database),
        new UserController(database),
        new CompositeAccessLog(new AccessLogController(database)),
        new RsaAuthenticator(),
        BCrypt::gensalt);
  }


```

That code block is effectively static and could be moved to a static factory somewhere else. *That* regardless of any other changes will remove the DB knowledge out from `LobbyLoginValidator`. 

The pattern of referencing a `this` constructor with dependencies injected is basically DI without IOC. IOC frameworks would make the above code 100% auto-magic (for better or worse). I've seen some people fake a spring way to provide beans by creating a static class that has a series of bean provider methods. I'm not necessarily a fan of that, IMO DI is best at the edges of modules for important strategies, and the above is good in strategic places to simplify the dependency stack.

My bottom line is then, the code above calling the `this` constructor is the one place that has the `Database` API dependency (or a `PropertyReader` + `new Database` dependency, either way). That code does not belong in this class to really begin with, it's static constructrion code. Because there is no IOC available to us, it's not the worst solution. Having `Database` as an API dependency in the effectively static constructor code only IMO is okay, as that code is not properly attached to `LobbyLoginValidator` anyways. In all other places, the proper instance methods of `LobbyLoginValidator`, there is no coupling to Database.

So long winded discussion, a potential other option *would* be to find a clean factory pattern for creating a `LobbyLoginValidator`, that would remove all of the "new'ing" of dependencies from `LobbyLoginValidator`.


Sorry for the lengthy text. We debated as a team to adopt an IOC and the decision was to avoid it unless really compelling. There are some problems I have trouble solving without an IOC, so in general I'm interested if we can come up with a clean solution. The maybe not great solution was to have `ClientContext` be our effective IOC. Going with that route, that would mean to get a `LobbyLoginValidator` you would get it from `ClientContext` (but since this is lobby contexct, hence, `LobbyContext`). So perhaps something to consider would be to restore the `LobbyContext` and put the above constructor code in `LobbyContext` and then grab `LobbyLoginValidator` instances from `LobbyContext`.

************

@ssoloff, for next steps, the how to design IOC without an IOC container is a larger conversation. I'd be okay with merging and incrementally updating/refactoring and discussing the bigger design questions in its own issue. We are *maybe* close to resolution on this topic here. I suspect if you see a good pattern in this existing code we can evaluate it and that would be the new de-factor best practice. By removing `LobbyContext` we are certainly making an implicit statement about the IOC direction we are taking.

Reply:
I agree there's plenty to improve with respect to dependency management within TripleA.  However, I don't think this PR is the place to tackle that.

The purpose of this PR is simply to refactor enough code to make it possible to test without relying on _lobby.properties_.  For quite some time before now, `LobbyLoginValidator` has been responsible for creating the DAOs and their dependents, and this PR doesn't change that.  If anything, it at least improves things a small bit by moving the knowledge of how to get a `Connection` factory from the DAOs to `LobbyLoginValidator`. So we lower the coupling in the DAOs while keeping the coupling within `LobbyLoginValidator` the same (because it already has a dependency on the DAO package internals).  Therefore, while DI in its current state is not ideal, we have "Done No Harm."